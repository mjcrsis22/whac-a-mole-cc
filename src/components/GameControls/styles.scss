.game-controls-container {
  padding: 0.5rem 0;

  .button {
    background-color: dodgerblue;
    color: white;
  }
}
