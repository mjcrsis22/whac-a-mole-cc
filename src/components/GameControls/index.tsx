import { useContext } from 'react';
import { GameContext } from '../../contexts/GameContext';
import { ActionTypes } from '../../contexts/GameContext/reducer';
import './styles.scss';

function GameControls() {
  const {
    state: { started },
    dispatch,
  } = useContext(GameContext);

  const toggleStarted = () => {
    dispatch({ type: ActionTypes.SET_STARTED, payload: !started });
  };

  return (
    <div className="game-controls-container">
      <button className="button" onClick={toggleStarted}>
        {started ? 'Stop' : 'Play'}
      </button>
    </div>
  );
}

export default GameControls;
