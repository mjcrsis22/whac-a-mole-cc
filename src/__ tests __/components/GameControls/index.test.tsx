import { useContext } from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import GameContextProvider from '../../../contexts/GameContext';
import {
  ActionTypes,
  INITIAL_STATE,
} from '../../../contexts/GameContext/reducer';
import GameControls from '../../../components/GameControls';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn(),
}));

const mockedContext = {
  state: INITIAL_STATE,
  dispatch: jest.fn(),
};

const mockedUseContext = useContext as jest.MockedFn<typeof useContext>;

describe('GameControls component', () => {
  beforeEach(() => {
    mockedUseContext.mockReturnValue(mockedContext);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test.each([true, false])(
    'Renders correctly - according to the started value',
    async started => {
      mockedUseContext.mockReturnValue({
        ...mockedContext,
        state: { ...mockedContext.state, started },
      });
      const { container } = render(
        <GameContextProvider>
          <GameControls />
        </GameContextProvider>,
      );
      expect(container).toMatchSnapshot();
    },
  );

  test('Dispatches SET_STARTED action on button click', async () => {
    const result = render(
      <GameContextProvider>
        <GameControls />
      </GameContextProvider>,
    );

    const button = result.getByRole('button');
    await userEvent.click(button);

    expect(result.container).toMatchSnapshot();
    expect(mockedContext.dispatch).toHaveBeenCalledWith({
      type: ActionTypes.SET_STARTED,
      payload: true,
    });
  });
});
