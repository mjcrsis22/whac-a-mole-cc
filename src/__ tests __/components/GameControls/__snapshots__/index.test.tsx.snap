// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameControls component Dispatches SET_STARTED action on button click 1`] = `
<div>
  <div
    class="game-controls-container"
  >
    <button
      class="button"
    >
      Play
    </button>
  </div>
</div>
`;

exports[`GameControls component Renders correctly - according to the started value 1`] = `
<div>
  <div
    class="game-controls-container"
  >
    <button
      class="button"
    >
      Stop
    </button>
  </div>
</div>
`;

exports[`GameControls component Renders correctly - according to the started value 2`] = `
<div>
  <div
    class="game-controls-container"
  >
    <button
      class="button"
    >
      Play
    </button>
  </div>
</div>
`;
